qe/drivers.py

```diff
@@ -1593,44 +1684,112 @@ def H(self, psi_internal: Psi_det, psi_external: Psi_det = None) -> List[List[En
+            S_k = np.r_[S_k, S_new_r]
+        # Diagonalize S_k and keep n_eig smallest estimates
+        L_k, Y_k = np.linalg.eigh(S_k)
+        L_k = L_k[:n_eig]
+        Y_k = Y_k[:, :n_eig]
+
+        n_newvecs = 0  # Initialize counter; no. of new vectors added to trial subspace
+        X_k = np.dot(V_k, Y_k)  # Pre-compute Ritz vectors (V_k updated each iteration)
+        converged = True  # Reset
+        for j in range(n_eig):
+            # Grab jth eigenpair, each rank computes local portion of jth residual vector
+            l_j, y_j = L_k[j], Y_k[:, j]
+            r_j = np.dot(W_k, y_j) - l_j * X_k[:, j]
+            res = np.linalg.norm(r_j)
+            if res > eps:  # If ||r_j|| > eps, jth eigenpair hasn't converged
+                converged = False
+                # Precondition next trial vector
+                M_j = np.diag(
+                    np.clip(np.reciprocal(D - l_j), a_min=-1e5, a_max=1e5)
+                )  # Build diagonal preconditioner
+                t_j = np.dot(M_j, r_j)
+                # Orthogonalize new trial vector against previous basis vectors using Modified Gram-Schmidt (MG)S
+                t_j = t_j / np.linalg.norm(t_j)
+                for i in range(V_k.shape[1]):  # Iterate through k basis vectors
+                    c_i = np.copy(
+                        np.inner(V_k[:, i], t_j)
+                    )  # Each process computes partial inner-product
+                    t_j -= c_i * V_k[:, i]  # Remove component of t_j in V_k
+
+                norm_tj = np.linalg.norm(t_j)
+                if norm_tj > 1e-4:  # If new vector is `small`, ignore. Avoids ill-conditioning
+                    V_k = np.c_[V_k, t_j / norm_tj]  # Append new vector to trial subspace
+                    n_newvecs += 1
+
+        if converged == True:
             break
-        theta_old = theta[:eig]
+
+        dim_S += n_newvecs  # Update dimension of trial subspace
+
+        if q <= dim_S:  # Collapose trial basis
+            V_new = np.c_[X_k[:, :n_eig], V_k[:, -n_newvecs:]]
+            # Take leading n_eig Ritz vectors as new guess vectors
+            dim_S = V_new.shape[1]  # New subspace dimension
+            n_newvecs = dim_S
+            W_k = np.zeros((self.local_size, 0), dtype="float")
+            V_k = np.linalg.qr(V_new)  # Brute-force a full QR
+            restart = 1  # Indicate restart # TODO: Cleaner way to do this?
+        elif n_newvecs == 0:
+            V_new = X_k[:, :n_eig]
+            # Take leading n_eig Ritz vectors as new guess vectors
+            dim_S = V_new.shape[1]  # New subspace dimension
+            n_newvecs = dim_S
+            W_k = np.zeros((self.local_size, 0), dtype="float")
+            V_k = np.linalg.qr(V_new)  # Brute-force a full QR
+            restart = 1  # Indicate restart
+
     else:
-        raise NotImplementedError("Not converged")
+        print(f"Max number of iterations reached. Returning {n_eig} lowest eigenvalue estimates")
```

Comment:
We should still raise, no point in continuing if we don't converge. The `max number of iterations` is here not to waste computation time on non converging hamiltonian 
Do you are @kgasperich ?